Gradle upgrade

ddv239 opened the pull request:
Upgrade Gradle, add annotations support and .editorconfig.

Adonai commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,7 +1,7 @@
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.1-all.zip
```
Hmmm, the README mentions Gradle 4.5 but wrapper downloads 4.1

ddv239 replied:
I am not sure how various components are versioned. I upgraded the Gradle "plugin" in the current version of Android Studio and it gave me this version. When I downloaded the current version of Graddle itself, to verify that I did not break the command-line build, it was 4.5, and I had do use a non-latest version of the JDK to make it work, so I added a note on that to readme in case someone else hits the same problem. I see no contradiction here.

adrian-bl replied:
Version mappings are listed here:
https://developer.android.com/studio/releases/gradle-plugin.html#updating-gradle

ddv239 replied:
OK, I can say that with this change, everything works perfectly correctly. Gradle 4.1 serves this change's purpose. The comment in Readme was about the "current version of Gradle" which is never than 4.1.

adrian-bl commented on .editorconfig:
```diff
@@ -0,0 +1,12 @@
+end_of_line=lf
+insert_final_newline=true
+indent_style=tab
+tab_width=4
```
This is not related to gradle, please revert this change.

ddv239 replied:
OK, will make a separate PR

adrian-bl replied:
I don't think we need to clutter the repo with this.
Every sane editor detects line-endings and tab/whitespace automatically anyway.

ddv239 replied:
I understand you but I have a sane Android Studio editor, and even though it automatically detects coding conventions already present in the sources and keeps my changes consistent with them, it pops a yellow warning bar on top of the text editor telling me that it replaced spaces with tabs to follow the document's formatting style but that does not match my editor preferences.
.editorconfig is the clean way to avoid any such inconsistencies and tell any sane editor what are the project's preferences that should override any globally configured user's preferences.
I truly believe you should not resist this clean and standard way to achieve this consistency and you have to admit that this "clutter" is truly minimal and so not worth avoiding.
By the way, presence of .editorconfig would even make it unnecessary to mention the corresponding formatting conventions in the Readme (which it currently does).

adrian-bl commented on README.md:
```diff
@@ -55,15 +55,17 @@ Building
 Building from command-line #2
 --------------------------
+> Note: at the time of this writing, the current version fo the Android SDK tools are not compatible with ant, see https://stackoverflow.com/questions/42912824/the-ant-folder-is-suddenly-missing-from-android-sdk-did-google-remove-it
  * `android update project --path .` to generate local.properties
```
We can remove this section as ant is not supported anymore.

ddv239 replied:
Sure

adrian-bl commented on README.md:
```diff
@@ -55,15 +55,17 @@ Building
 To build you will need:
 
  * A Java compiler compatible with Java 1.7
```
s/1.7/1.8/

ddv239 replied:
I don't understand what this comment means

adrian-bl replied:
Replace 1.7 with 1.8, as the text below (correctly) states that java >= 1.8 is required.

ddv239 replied:
Actually, you already replaced that in a previous change so my updated version of this PR will not have this line changed.

adrian-bl commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,7 +1,7 @@
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.1-all.zip
+distributionSha256Sum=71a787faed83c4ef21e8464cc8452b941b5fcd575043aa29d39d15d879be89f7
```
This sha256 sum is wrong.

ddv239 replied:
Good catch; I just had android Studio make this change for me and did not realize it failed to properly update the checksum. Will make sure the updated version does contain a valid one.

adrian-bl commented on build.gradle:
```diff
@@ -52,3 +53,7 @@ android {
 }
+
+dependencies {
+    compileOnly  'com.android.support:support-annotations:27.0.2'
```
I still don't get what exactly we get from this.
Is it an Android-Studio thing?

ddv239 replied:
As mentioned before in a comment to #716: it's just annotations part of it for making code look cleaner (e.g. @NonNull or @Nullable etc. JSR-305). It is not specific to Android Studio and is used during compilation.

ddv239 commented:
I have pushed a new state to this branch which isolates the "Gradle upgrade" only, with no annotations support or .editorconfig additions, addressing all the comments.
